core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.ListMultimap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Multimaps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.StructLikeWrapper;
+import org.apache.iceberg.util.Tasks;
+
+/**
+ * An index of {@link DeleteFile delete files} by sequence number.
+ * <p>
+ * Use {@link #builderFor(FileIO, Iterable)} to construct an index, and {@link #forDataFile(int, long, DataFile)} or
+ * {@link #forEntry(int, ManifestEntry)} to get the the delete files to apply to a given data file.
+ */
+class DeleteFileIndex {
+  private static final DeleteFile[] NO_DELETE_FILES = new DeleteFile[0];
+
+  private final long[] globalSeqs;
+  private final DeleteFile[] globalDeletes;
+  private final Map<Pair<Integer, StructLikeWrapper>, Pair<long[], DeleteFile[]>> sortedDeletesByPartition;
+  private final ThreadLocal<StructLikeWrapper> lookupWrapper = ThreadLocal.withInitial(
+      () -> StructLikeWrapper.wrap(null));
+
+  DeleteFileIndex(long[] globalSeqs, DeleteFile[] globalDeletes,
+                  Map<Pair<Integer, StructLikeWrapper>, Pair<long[], DeleteFile[]>> sortedDeletesByPartition) {
+    this.globalSeqs = globalSeqs;
+    this.globalDeletes = globalDeletes;
+    this.sortedDeletesByPartition = sortedDeletesByPartition;
+  }
+
+  DeleteFile[] forEntry(int specId, ManifestEntry<DataFile> entry) {
+    return forDataFile(specId, entry.sequenceNumber(), entry.file());
+  }
+
+  DeleteFile[] forDataFile(int specId, long sequenceNumber, DataFile file) {
+    Pair<long[], DeleteFile[]> partitionDeletes = sortedDeletesByPartition
+        .get(Pair.of(specId, lookupWrapper.get().set(file.partition())));
+
+    if (partitionDeletes == null) {
+      return limitBySequenceNumber(sequenceNumber, globalSeqs, globalDeletes);
+    } else if (globalDeletes == null) {
+      return limitBySequenceNumber(sequenceNumber, partitionDeletes.first(), partitionDeletes.second());
+    } else {
+      return Stream.concat(
+          Stream.of(limitBySequenceNumber(sequenceNumber, globalSeqs, globalDeletes)),
+          Stream.of(limitBySequenceNumber(sequenceNumber, partitionDeletes.first(), partitionDeletes.second()))
+      ).toArray(DeleteFile[]::new);
+    }
+  }
+
+  private static DeleteFile[] limitBySequenceNumber(long sequenceNumber, long[] seqs, DeleteFile[] files) {
+    if (files == null) {
+      return NO_DELETE_FILES;
+    }
+
+    int pos = Arrays.binarySearch(seqs, sequenceNumber);
+    int start;
+    if (pos < 0) {
+      // the sequence number was not found, where it would be inserted is -(pos + 1)
+      start = -(pos + 1);
```

Comment:
Does this cover cases when all delete files or none match? I think `Arrays.copyOfRange` will throw an exception if `from > original.length`.

Reply:
Yes, and the tests validate these cases.

If the sequence number is not found and less than all of the sequence numbers in the array, then the insert position is 0 and the `pos` returned is `-(0 + 1) = -1`. Converting back to a start position `-(-1 + 1) = 0`, so we copy the entire array. Similarly, if the sequence number is greater than all of the numbers in the array, the return values is `length`, which results in `copyOfRange(files, length, length)` and produces a 0-length array.

Reply:
> Yes, and the tests validate these cases.
> 
> If the sequence number is not found and less than all of the sequence numbers in the array, then the insert position is 0 and the `pos` returned is `-(0 + 1) = -1`. Converting back to a start position `-(-1 + 1) = 0`, so we copy the entire array. Similarly, if the sequence number is greater than all of the numbers in the array, the return values is `length`, which results in `copyOfRange(files, length, length)` and produces a 0-length array.

naive question, since not found returns -1, why not just set start to 0 ?